pkg/kubectl/rolling_updater.go

```diff
@@ -93,27 +96,458 @@ const (
+// rollout.
+func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
+	out := config.Out
+	oldRc := config.OldRc
+	scaleRetryParams := NewRetryParams(config.Interval, config.Timeout)
+
+	// Find an existing controller (for continuing an interrupted update) or
+	// create a new one if necessary.
+	sourceId := fmt.Sprintf("%s:%s", oldRc.Name, oldRc.UID)
+	newRc, existed, err := r.getOrCreateTargetController(config.NewRc, sourceId)
+	if err != nil {
+		return err
+	}
+	if existed {
+		fmt.Fprintf(out, "Continuing update with existing controller %s.\n", newRc.Name)
+	} else {
+		fmt.Fprintf(out, "Created %s\n", newRc.Name)
+	}
+	// Extract the desired replica count from the controller.
+	desired, err := strconv.Atoi(newRc.Annotations[desiredReplicasAnnotation])
+	if err != nil {
+		return fmt.Errorf("Unable to parse annotation for %s: %s=%s",
+			newRc.Name, desiredReplicasAnnotation, newRc.Annotations[desiredReplicasAnnotation])
+	}
+	// Extract the original replica count from the old controller, adding the
+	// annotation if it doesn't yet exist.
+	_, hasOriginalAnnotation := oldRc.Annotations[originalReplicasAnnotation]
+	if !hasOriginalAnnotation {
+		existing, err := r.c.ReplicationControllers(oldRc.Namespace).Get(oldRc.Name)
+		if err != nil {
+			return err
+		}
+		if existing.Annotations == nil {
+			existing.Annotations = map[string]string{}
+		}
+		existing.Annotations[originalReplicasAnnotation] = strconv.Itoa(existing.Spec.Replicas)
+		updated, err := r.c.ReplicationControllers(existing.Namespace).Update(existing)
+		if err != nil {
+			return err
+		}
+		oldRc = updated
+	}
+	original, err := strconv.Atoi(oldRc.Annotations[originalReplicasAnnotation])
+	if err != nil {
+		return fmt.Errorf("Unable to parse annotation for %s: %s=%s\n",
+			oldRc.Name, originalReplicasAnnotation, oldRc.Annotations[originalReplicasAnnotation])
+	}
+
+	// The maximum pods which can go unavailable during the update.
+	maxUnavailable, err := extractMaxValue(config.MaxUnavailable, "maxUnavailable", original)
+	if err != nil {
+		return err
+	}
+	// The maximum scaling increment.
+	maxSurge, err := extractMaxValue(config.MaxSurge, "maxSurge", original)
+	if err != nil {
+		return err
+	}
+	// Further validation.
+	if maxUnavailable == 0 && maxSurge == 0 {
+		return fmt.Errorf("one of maxSurge or maxUnavailable must be specified")
+	}
+	// The minumum pods which must remain available througout the update
+	// calculated for internal convenience.
+	minAvailable := original - maxUnavailable
+	// Set the initial increment. This will drive the direction of the update;
+	// an initial increment of 0 means the first scale-up will be skipped,
+	// causing a down/up direction.
+	increment := 0
+	// A surge value means we need to move up/down to make availability for a
+	// scale down, and we'll burst higher than if only a maxUnavailable were
+	// set.
+	if maxSurge > 0 {
+		increment = maxUnavailable + maxSurge
```

Comment:
The fact that it's okay for the scale up to happen first here solves my
problem. Thanks for your help, guys. I'll have some much simpler code for
review very soon.

On Tue, Aug 25, 2015 at 2:01 PM, Brian Grant <notifications@github.com>
wrote:

> In pkg/kubectl/rolling_updater.go
> <https://github.com/kubernetes/kubernetes/pull/11942#discussion_r37897537>
> :
>
> > +	// Further validation.
> > +	if maxUnavailable == 0 && maxSurge == 0 {
> > +		return fmt.Errorf("one of maxSurge or maxUnavailable must be specified")
> > +	}
> > +	// The minumum pods which must remain available througout the update
> > +	// calculated for internal convenience.
> > +	minAvailable := original - maxUnavailable
> > +	// Set the initial increment. This will drive the direction of the update;
> > +	// an initial increment of 0 means the first scale-up will be skipped,
> > +	// causing a down/up direction.
> > +	increment := 0
> > +	// A surge value means we need to move up/down to make availability for a
> > +	// scale down, and we'll burst higher than if only a maxUnavailable were
> > +	// set.
> > +	if maxSurge > 0 {
> > +		increment = maxUnavailable + maxSurge
>
> Sorry, my previous example was buggy. We could scale down the original RC
> by 10% and then create the new RC at 30%, or create the new RC at 20%
> before scaling down the original RC by 10%. The user might prefer the
> latter, and it might also be simpler to implement.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/11942/files#r37897537>.
>